In [102]:
from autenticazione.autenticazione import start_kaggle
start_kaggle()


In [103]:
# Secondo passo: Download del dataset
# Visualizzare la lista dei dataset
import kaggle

key_search = "dog"

datasets = kaggle.api.datasets_list(search=key_search)

In [104]:
# stampo la lista dei dataset
# la lista di dataset è una lista di dizionari, il cui valore di owner e' contenuto nella key 'ref'
for dataset in datasets:
    print(dataset['ref'])

tongpython/cat-and-dog
dansbecker/hot-dog-not-hot-dog
jessicali9530/stanford-dogs-dataset
marshuu/dog-breeds
shaunthesheep/microsoft-catsvsdogs-dataset
andrewmvd/dog-and-cat-detection
danielshanbalico/dog-emotion
whenamancodes/dog-adoption
chetankv/dogs-cats-images
dhruvildave/dogecoin-historical-data
mmoreaux/audio-cats-and-dogs
devzohaib/dog-emotions-prediction
kingburrito666/largest-dog-breed-data-set
d4rklucif3r/cat-and-dogs
arpitjain007/dog-vs-cat-fastai
mohamedchahed/dog-breeds
new-york-city/nyc-dog-names
gpiosenka/70-dog-breedsimage-data-set
kmader/dogs-of-zurich
biaiscience/dogs-vs-cats


In [105]:
# scelgo il dataset che mi interessa e lo scarico
# imposto il path in cui scaricare il dataset
# path = os.path.dirname(os.path.abspath(__file__)) + '/dataset' su juptyer notebook non funziona
# usare invece sys.path[0] + '/dataset' con la libreria sys
import sys
path = sys.path[0] + '/dataset'

# scelgo il dataset che mi interessa
dt_scelto = 'marshuu/dog-breeds'

# scarico il dataset
kaggle.api.dataset_download_files(dt_scelto, path=path, unzip=True)

In [106]:
# ispeziono il dataset
# creo un dataframe con il dataset
import pandas as pd
df = pd.read_csv(path + '/dog_breeds.csv')

ANALIZIAMO IL DATAFRAME

In [107]:
# visualizzo le prime 5 righe del dataframe
df.head()

Breed Country of Origin                     Fur Color  \
0  Labrador Retriever            Canada      Yellow, Black, Chocolate   
1     German Shepherd           Germany                    Black, Tan   
2             Bulldog           England                    White, Red   
3              Poodle            France  White, Black, Brown, Apricot   
4              Beagle           England        White, Tan, Red, Lemon   

  Height (in) Color of Eyes Longevity (yrs)  \
0       21-24         Brown           10-12   
1       22-26         Brown            7-10   
2       12-16         Brown            8-10   
3       10-15   Brown, Blue           12-15   
4       13-15         Brown           12-15   

                                    Character Traits  \
0  Loyal, friendly, intelligent, energetic, good-...   
1  Loyal, intelligent, protective, confident, tra...   
2                         Loyal, calm, gentle, brave   
3  Intelligent, active, affectionate, hypoallergenic   
4         Curious, friendly, energetic, good-natured   

                         Common Health Problems  
0        Hip dysplasia, obesity, ear infections  
1  Hip dysplasia, elbow dysplasia, pancreatitis  
2   Skin allergies, respiratory issues, obesity  
3       Hip dysplasia, epilepsy, bladder stones  
4       Ear infections, hip dysplasia, epilepsy

In [108]:
# analiziamo la struttura del dataframe
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 117 entries, 0 to 116
Data columns (total 8 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   Breed                   117 non-null    object
 1   Country of Origin       117 non-null    object
 2   Fur Color               117 non-null    object
 3   Height (in)             117 non-null    object
 4   Color of Eyes           117 non-null    object
 5   Longevity (yrs)         117 non-null    object
 6   Character Traits        117 non-null    object
 7   Common Health Problems  117 non-null    object
dtypes: object(8)
memory usage: 7.4+ KB


Il dataframe si compone di 117 rows e 7 colonne. Non ci sono elementi nulli e tutti i valori sono degli oggetti di tipo stringa.

In [109]:
# analizziamo le statistiche del dataframe
df.describe()

Breed Country of Origin Fur Color Height (in)  \
count                   117               117       117         117   
unique                  103                27        69          67   
top     Australian Shepherd           England     White       10-12   
freq                      3                24         9           8   

       Color of Eyes Longevity (yrs)  \
count            117             117   
unique             4              15   
top            Brown           12-15   
freq             108              44   

                                     Character Traits  \
count                                             117   
unique                                             38   
top     Intelligent, energetic, playful, good-natured   
freq                                               50   

                             Common Health Problems  
count                                           117  
unique                                           30  
top     Dental problems, eye issues, skin allergies  
freq                                             59

Il dataframe presente 103 razze di cane uniche elencate e 27 paesi di origine. I colori sono 69.

In [110]:
# trasformiamo la colonna 'Longevity (yrs)' in una colonna numerica divisa in due colonne 'Longevity (yrs) min' e 'Longevity (yrs) max'
# per fare questo, creiamo una funzione che prende in input una stringa e restituisce una lista di due elementi
# la funzione prende in input una stringa del tipo '12-15' e restituisce una lista di due elementi [12, 15]
def split_string(stringa):
    lista = stringa.split('-')
    lista = [int(x) for x in lista]
    return lista

# applichiamo la funzione alla colonna 'Longevity (yrs)'
# usiamo il metodo apply() di pandas per applicare una funzione alla colonna 'Longevity (yrs)'
df['Longevity (yrs)'] = df['Longevity (yrs)'].apply(split_string)

# creiamo le due nuove colonne 'Longevity (yrs) min' e 'Longevity (yrs) max'
df['Longevity (yrs) min'] = df['Longevity (yrs)'].apply(lambda x: x[0])     # la funzione lambda prende in input una lista e restituisce il primo elemento
df['Longevity (yrs) max'] = df['Longevity (yrs)'].apply(lambda x: x[1])     # la funzione lambda prende in input una lista e restituisce il secondo elemento

# ora possiamo eliminare la colonna 'Longevity (yrs)'
df.drop('Longevity (yrs)', axis=1, inplace=True)  # potremmo anche usare del df['Longevity (yrs)']

In [ ]:
df.head()   

In [111]:
# effettuiamo la stessa cosa per la colonna 'Height (cm)'
df['Height (in)'] = df['Height (in)'].apply(split_string)
df['Height (in) min'] = df['Height (in)'].apply(lambda x: x[0])
df['Height (in) max'] = df['Height (in)'].apply(lambda x: x[1])
df.drop('Height (in)', axis=1, inplace=True)

In [112]:
# analiziamo i dati della colonna 'Character Traits' e creiamo una colonna per ogni caratteristica
# la colonna 'Character Traits' contiene una lista di caratteristiche
# creiamo una funzione che prende in input una lista e restituisce una lista di caratteristiche
def split_string_2(stringa):
    lista = stringa.split(',')
    lista = [x.strip() for x in lista]
    return lista

# applichiamo la funzione alla colonna 'Character Traits'
df['Character Traits'] = df['Character Traits'].apply(split_string_2)

# creiamo una lista di caratteristiche
lista_caratteristiche = []
for lista in df['Character Traits']:
    for caratteristica in lista:
        if caratteristica not in lista_caratteristiche:
            lista_caratteristiche.append(str(caratteristica).lower())  

# conserviamo solo i valori unici della lista
import numpy as np
lista_caratteristiche = np.unique(lista_caratteristiche)

# creiamo una colonna per ogni caratteristica
for caratteristica in lista_caratteristiche:
    df[caratteristica] = df['Character Traits'].apply(lambda x: 1 if caratteristica in str(x).lower() else 0)


In [113]:
print(lista_caratteristiche)

['active' 'affectionate' 'athletic' 'brave' 'calm' 'charming' 'confident'
 'curious' 'energetic' 'friendly' 'gentle' 'good-natured' 'hypoallergenic'
 'independent' 'intelligent' 'kind' 'loyal' 'patient' 'playful'
 'protective' 'sensitive' 'social' 'strong' 'trainable']


In [114]:
# eliminiamo la colonna 'Character Traits'
df.drop('Character Traits', axis=1, inplace=True)

In [115]:
# applichiamo la funzione anche alla colonna 'Common Health Problems'
df['Common Health Problems'] = df['Common Health Problems'].apply(split_string_2)

# creiamo una lista di problemi di salute
lista_problemi_salute = []
for lista in df['Common Health Problems']:
    for problema_salute in lista:
        if problema_salute not in lista_problemi_salute:
            lista_problemi_salute.append(problema_salute)

# conserviamo solo i valori unici della lista
lista_problemi_salute = np.unique(lista_problemi_salute)

# creiamo una colonna per ogni problema di salute
for problema_salute in lista_problemi_salute:
    df[problema_salute] = df['Common Health Problems'].apply(lambda x: 1 if problema_salute in x else 0)

# eliminiamo la colonna 'Common Health Problems'
df.drop('Common Health Problems', axis=1, inplace=True)

In [116]:
# vediamo quale razza di cane ha la vita media piu' lunga
df.sort_values(by='Longevity (yrs) max', ascending=False).head()

Breed Country of Origin                 Fur Color Color of Eyes  \
5         Chihuahua            Mexico  Black, Brown, Tan, White   Brown, Blue   
46  Chinese Crested             China        Pink, White, Black         Brown   
18       Pomeranian           Germany      Orange, Sable, Black         Brown   
51   Border Terrier           England       Red, Wheaten, Black         Brown   
37         Papillon            France  White, Black, Sable, Red         Brown   

    Longevity (yrs) min  Longevity (yrs) max  Height (in) min  \
5                    12                   20                6   
46                   13                   18               11   
18                   12                   16                6   
51                   12                   16               10   
37                   14                   16                8   

    Height (in) max  active  affectionate  ...  epilepsy  eye issues  \
5                 9       0             0  ...         0           1   
46               13       0             1  ...         0           0   
18                7       0             0  ...         0           0   
51               16       0             0  ...         0           0   
37               11       0             1  ...         0           0   

    eye problems  heart conditions  hereditary myopathy  hip dysplasia  \
5              0                 0                    0              0   
46             1                 0                    0              0   
18             1                 0                    0              0   
51             0                 0                    0              1   
37             1                 0                    0              0   

    obesity  pancreatitis  respiratory issues  skin allergies  
5         0             0                   1               0  
46        0             0                   0               1  
18        1             0                   0               0  
51        0             0                   0               1  
37        0             0                   0               0  

[5 rows x 54 columns]

In [117]:
# vediamo in quale paese si trovano le razze di cani con la vita media piu' lunga
df.groupby('Country of Origin')['Longevity (yrs) max'].mean().sort_values(ascending=False)

Country of Origin
Mexico           17.500000
China            15.166667
Africa           15.000000
Tibet            15.000000
Netherlands      15.000000
Congo            15.000000
Finland          15.000000
Australia        14.666667
Malta            14.500000
Wales            14.500000
Belgium          14.250000
Japan            14.000000
Madagascar       14.000000
Middle East      14.000000
Turkey           14.000000
France           13.900000
Scotland         13.888889
England          13.708333
Russia           13.666667
United States    13.600000
Italy            13.500000
Hungary          13.000000
Germany          12.923077
Ireland          12.714286
Rhodesia         12.000000
Canada           11.500000
Switzerland      10.000000
Name: Longevity (yrs) max, dtype: float64

In [118]:
# vediamo quale cane ha l'altezza media piu' alta
df.sort_values(by='Height (in) max', ascending=False).head()

Breed Country of Origin                     Fur Color  \
24  Irish Wolfhound           Ireland              Grey, Red, Black   
86  Irish Wolfhound           Ireland                     Grey, Red   
20          Mastiff           England                 Apricot, Fawn   
19       Great Dane           Germany  Black, Blue, Fawn, Harlequin   
27   Great Pyrenees            France                         White   

   Color of Eyes  Longevity (yrs) min  Longevity (yrs) max  Height (in) min  \
24         Brown                    6                    8               30   
86         Brown                    6                   10               30   
20         Brown                    8                   10               30   
19         Brown                    8                   10               28   
27         Brown                   10                   12               25   

    Height (in) max  active  affectionate  ...  epilepsy  eye issues  \
24               35       0             0  ...         0           0   
86               35       0             0  ...         0           1   
20               34       0             0  ...         0           0   
19               34       0             0  ...         0           0   
27               32       0             0  ...         0           0   

    eye problems  heart conditions  hereditary myopathy  hip dysplasia  \
24             0                 1                    0              0   
86             0                 0                    0              0   
20             0                 1                    0              0   
19             0                 1                    0              0   
27             0                 0                    0              0   

    obesity  pancreatitis  respiratory issues  skin allergies  
24        0             0                   0               0  
86        0             0                   0               1  
20        0             0                   0               0  
19        0             0                   0               0  
27        0             0                   0               1  

[5 rows x 54 columns]

In [119]:
# raggruppando per razza vediamo quale cane ha l'altezza media piu' alta
df.groupby('Breed')['Height (in) max'].mean().sort_values(ascending=False)

Breed
Irish Wolfhound       35.0
Great Dane            34.0
Mastiff               34.0
Great Pyrenees        32.0
Scottish Deerhound    32.0
                      ... 
Pekingese              9.0
Dachshund              9.0
Chihuahua              9.0
Yorkshire Terrier      8.0
Pomeranian             7.0
Name: Height (in) max, Length: 103, dtype: float64

In [120]:
# raggruppiamo per razza e vediamo la causa di morte piu' comune
df.groupby('Breed')[lista_problemi_salute].sum()

Breathing problems  Dental problems  \
Breed                                                              
Affenpinscher                                 0                1   
Africanis                                     0                1   
Akita                                         0                1   
American Bulldog                              0                1   
American Eskimo Dog                           0                1   
...                                         ...              ...   
West Highland White Terrier                   0                2   
Whippet                                       0                1   
Wirehaired Pointing Griffon                   0                1   
Xoloitzcuintli                                0                1   
Yorkshire Terrier                             0                1   

                             Ear infections  Eye problems  Hip dysplasia  \
Breed                                                                      
Affenpinscher                             0             0              0   
Africanis                                 0             0              0   
Akita                                     0             0              0   
American Bulldog                          0             0              0   
American Eskimo Dog                       0             0              0   
...                                     ...           ...            ...   
West Highland White Terrier               0             0              0   
Whippet                                   0             0              0   
Wirehaired Pointing Griffon               0             0              0   
Xoloitzcuintli                            0             0              0   
Yorkshire Terrier                         0             0              0   

                             Intervertebral disc disease  Skin allergies  \
Breed                                                                      
Affenpinscher                                          0               0   
Africanis                                              0               0   
Akita                                                  0               0   
American Bulldog                                       0               0   
American Eskimo Dog                                    0               0   
...                                                  ...             ...   
West Highland White Terrier                            0               0   
Whippet                                                0               0   
Wirehaired Pointing Griffon                            0               0   
Xoloitzcuintli                                         0               0   
Yorkshire Terrier                                      0               0   

                             bladder stones  cancer  diabetes  ...  epilepsy  \
Breed                                                          ...             
Affenpinscher                             0       0         0  ...         0   
Africanis                                 0       0         0  ...         0   
Akita                                     0       0         0  ...         0   
American Bulldog                          0       0         0  ...         0   
American Eskimo Dog                       0       0         0  ...         0   
...                                     ...     ...       ...  ...       ...   
West Highland White Terrier               0       0         0  ...         0   
Whippet                                   0       0         0  ...         0   
Wirehaired Pointing Griffon               0       0         0  ...         0   
Xoloitzcuintli                            0       0         0  ...         0   
Yorkshire Terrier                         0       0         0  ...         0   

                             eye issues  eye problems  heart conditions  \
Breed                                             

In [121]:
# vediamo quale razza ha piu' problemi di salute
# sommiamo il numero di problemi di salute per ogni razza
df.groupby('Breed')[lista_problemi_salute].sum().sum(axis=1).sort_values(ascending=False)

Breed
Australian Shepherd         9
Gordon Setter               6
Pug                         6
English Setter              6
Golden Retriever            6
                           ..
Chinese Crested             3
Chihuahua                   3
Chesapeake Bay Retriever    3
Catahoula Leopard Dog       3
Yorkshire Terrier           3
Length: 103, dtype: int64

In [ ]:
from libretranslatepy import LibreTranslateAPI

lt = LibreTranslateAPI("https://translate.argosopentech.com/")

In [122]:
# creiamo una funzione di traduzione che prende in input una stringa e restituisce la traduzione in italiano
def traduci(stringa):
    # individua la lingua della stringa
    lingua = lt.detect(stringa)[0]['language']
    # traduci la stringa in italiano
    stringa_tradotta = lt.translate(stringa, lingua, 'it')
    return stringa_tradotta


In [123]:
print(lista_caratteristiche)

['active' 'affectionate' 'athletic' 'brave' 'calm' 'charming' 'confident'
 'curious' 'energetic' 'friendly' 'gentle' 'good-natured' 'hypoallergenic'
 'independent' 'intelligent' 'kind' 'loyal' 'patient' 'playful'
 'protective' 'sensitive' 'social' 'strong' 'trainable']


In [124]:
for i in range(len(lista_caratteristiche)):
    print(traduci(lista_caratteristiche[i]))

attivo
affettuoso
atletica atletica
coraggioso
calma
affascinante
fiducioso
curiosi
energia
amichevole
gentile
buona natura
Ipoallergenico
indipendente
intelligente intelligente
tipo
leale leale
paziente
giocoso
protezione
sensibili
sociale
forte
trainabile


In [125]:
dizionario_caratteristiche = {}
for caratteristica in lista_caratteristiche:
    print(traduci(caratteristica))
    importanza = input('Inserisci il valore di importanza (1 o -1): ')
    dizionario_caratteristiche[caratteristica] = int(importanza)

attivo
affettuoso
atletica atletica
coraggioso
calma
affascinante
fiducioso
curiosi
energia
amichevole
gentile
buona natura
Ipoallergenico
indipendente
intelligente intelligente
tipo
leale leale
paziente
giocoso
protezione
sensibili
sociale
forte
trainabile


In [126]:
df.head()   

Breed Country of Origin                     Fur Color  \
0  Labrador Retriever            Canada      Yellow, Black, Chocolate   
1     German Shepherd           Germany                    Black, Tan   
2             Bulldog           England                    White, Red   
3              Poodle            France  White, Black, Brown, Apricot   
4              Beagle           England        White, Tan, Red, Lemon   

  Color of Eyes  Longevity (yrs) min  Longevity (yrs) max  Height (in) min  \
0         Brown                   10                   12               21   
1         Brown                    7                   10               22   
2         Brown                    8                   10               12   
3   Brown, Blue                   12                   15               10   
4         Brown                   12                   15               13   

   Height (in) max  active  affectionate  ...  epilepsy  eye issues  \
0               24       0             0  ...         0           0   
1               26       0             0  ...         0           0   
2               16       0             0  ...         0           0   
3               15       1             1  ...         1           0   
4               15       0             0  ...         1           0   

   eye problems  heart conditions  hereditary myopathy  hip dysplasia  \
0             0                 0                    0              0   
1             0                 0                    0              0   
2             0                 0                    0              0   
3             0                 0                    0              0   
4             0                 0                    0              1   

   obesity  pancreatitis  respiratory issues  skin allergies  
0        1             0                   0               0  
1        0             1                   0               0  
2        1             0                   1               0  
3        0             0                   0               0  
4        0             0                   0               0  

[5 rows x 54 columns]

In [127]:
# raggruppiamo per razza e vediamo chi ha piu' caratteristiche positive
df.groupby('Breed')[lista_caratteristiche].sum().sum(axis=1).sort_values(ascending=False)


Breed
Australian Shepherd         12
Pug                          9
Golden Retriever             9
Gordon Setter                8
Cocker Spaniel               8
                            ..
Chinese Crested              4
Chihuahua                    4
Chesapeake Bay Retriever     4
Yorkshire Terrier            4
French Bulldog               3
Length: 103, dtype: int64